### PR TITLE
Circuitry 2.0

### PR DESCRIPTION
**Pull request by PaulRitter**
[wip]
putting the pics on top cause that what everyone wants to see anyways (wall of text is below)
bMBOatahSo

Current State:
system works, now adding new content/features

Features:
- Ontouch events for every components, can be toggled off
- Assemblies can have a fixed size
- Assemblies can be attached to/detached from ANY object
- Assemblies can access attached object to work their magic
- Assemblies can be fully editied, this means
-- Removing/Adding Components
-- Connecting/Disconnecting Components
- Outputs get fired by a subsystem to prevent recursion

Components:
- [X] Doorcontroller (with setable access)
- [X] Signaler (receives and sends)
- [X] Button (also a toggle version)
- [X] Debugger
- [X] Splitter
  - [ ] SettingsUI
- [ ] Radio (can install encryptionkeys, kinda mimics the headset)
- [x] Speaker
- [ ] Sound Synth
- [ ] Movement Component (would only work on raw assemblies or specific objects)
- [x] Keyboard
- [x] Proximity Sensor
- [x] algorithmic components eg. 1+2 = 4
- [x] cleaner (cleaners tile undeneath), so you can build a screaming roomba
- [ ] manipulator (be ready for doors slapping you if you try to open them with the wrong access) (or a knifewielding roomba pointing at you only to then start charging at you while playing his integrated guitar)
  - [ ] grabbing, using things
  - [ ] pointing
  - [ ] harm intent if emagged
- [ ] Door reader (handles callbacks like denied(), open(), close(), might add this to doorcontroller)
- [ ] tracking beacon, gps getter, suit sensors
- [ ] eyes
- logic components eg. check if signal is the type of a specific obj, check if signal is 15, ...
  - [x] typecheck, would also include isnum/istext
  - [ ] equals check
  - [ ] not
- [ ] some sort of cargo component
- [ ] Accesschecker
- [x] String appender (appends something to the signal)
- [ ] timer
- [x] list operators
  - [x] index getter
  - [x] list iterator
- [ ] medical scanner
- [ ] pai component

Assemblytypes:
- [X] Base
- [ ] Robot (accepts all components, can't be attached) -- different sizes
- [ ] attachable (accepts many components, can be attached) //eg. no movement, altough maybe manipulation -- different sizes
- [ ] handheld (fits in pocket, accepts all components attachable does)
- [ ] tablet/screen (basically handheld but doesnt fit in a pocket, also can take components)
- [ ] anchored assembly that can take ALOT of components

OTHER TODO
- [x] add a debug input trigger button to the ui
- [x] add a button to erase a set output instead of setting it new all the time
- [x] add component proc to fire when emagged that changes behaviour
- [ ] make the editing ui not shit
- [ ] write a somehow decent tutorial on how to add your own components/make an example component
- [x] component whitelist and/or blacklist for assemblies
- [ ] an assembly locking mechanism
- [ ] a way to create assemblies and components
- [ ] a way to save assemblies and print them quickly (but expensively)
- [x] prevent endless loops?
- [ ] someway to show attached assemblies on an obj, maybe even somewhat dynamically
- [ ] sprites, please @ me in discord if you wanna help. i got no talent
  - overlay system for overlays like the manipulator, so i can use inhand sprites
- more chat-feedback
  - [ ] component cant be added to assembly
- [ ] test attachment system (flags)

BUGS
- [ ] for some reason handleOutput gets called again after executing the output, altough with an invalid vgc so it gets squashed by sanitychecks in fireOutputs(), resulting in only one output per outputcall
- [ ] components with no inputs show up on inputselect

### CODE REVIEW

**Comment by Exxion** [comment]
Just so you know, it's spelled "circuit"

**Comment by PaulRitter** [comment]
am german have excuse

also thank you

File: code/modules/vgcomponents/vgc.dm [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        return //how
+
+    _output[out] = list(vgc, target)
+
+/*
+Door control
+-- maybe let this send out events sometime like ondooropen, ondoorclose
+*/
+datum/vgcomponent/doorController
+    var/list/saved_access = null //ID.GetAccess()
```
Comment (ShiftyRail) [inline code/modules/vgcomponents/vgc.dm]: `= list()` is equivalent and less likely to runtime

File: code/modules/vgcomponents/vgc.dm [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+*/
+/datum/vgcomponent/debugger
+	var/spam = 1
+
+datum/vgcomponent/doorController/getPhysical()
+    return new obj/item/vgc_obj/debugger(src)
+
+/datum/vgcomponent/debugger/main(var/signal)
+	if(spam)
+		to_chat(world, "received signal:[signal] | <a HREF='?src=\ref[src];pause=1'>\[Toggle Output\]</a>")
```
Comment (ShiftyRail) [inline code/modules/vgcomponents/vgc.dm]: Hitler

File: code/modules/vgcomponents/vgc.dm (hunk 1 above)
Comment (PaulRitter) [reply code/modules/vgcomponents/vgc.dm]: tru

File: code/modules/vgcomponents/vgc.dm (hunk 2 above)
Comment (PaulRitter) [reply code/modules/vgcomponents/vgc.dm]: ~i dont get it~ its debug reeeeee

Comment (ShiftyRail) [reply code/modules/vgcomponents/vgc.dm]: `to_chat(world)`s are refered as "Hitlers" and should be avoided, even as debug messages
Make it a `message_admins` or whatever

**Comment by PaulRitter** [comment]
yes i know exxion its spelled incosistent

~im just to lazy to rename my commit~
~and totally didn't google it~
~and realized that i fucked up~
so as it turns out compiling is really boring

**Comment by Kurfursten** [comment]
You may request to have this reopened

**Comment by PaulRitter** [comment]
i fucked up merging the updated branch, will retry

**Comment by SonixApache** [comment]
medical scanner / suit sensors

**Comment by PaulRitter** [comment]
its going derelict again, sorry.

i'll tell you when/if i ever pick this up again

**Pull request closed without merging**